Python3.13 deprecated attribute "_acquireLock" from logging module

***In GitLab by @LudoBroche on Jan 22, 2025, 18:43 GMT+1:***

"_acquireLock" from the logging module is deprecated in python3.13
[https://redirect.github.com/python/cpython/issues/105182](https://redirect.github.com/python/cpython/issues/105182)


Error: in ewokscore/events/global_state.py
```
  def _ewoks_event_logger(
        handlers: Optional[List[Dict[str, str]]] = None, asynchronous: Optional[bool] = None
    ) -> Iterable[logging.Logger]:
        """Initialize and yield the EWOKS event logger"""
        # Issue with logging and forking:
        # https://pythonspeed.com/articles/python-multiprocessing/
    
>       logging._acquireLock()
E       AttributeError: module 'logging' has no attribute '_acquireLock'
/usr/local/lib/python3.13/site-packages/ewokscore/events/global_state.py:87: AttributeError
```



How to reproduce: https://gitlab.esrf.fr/workflow/ewoksapps/ewoksid29.git
with test-3.13_glx CI.

**Assignees:** @woutdenolf

*Migrated from GitLab: https://gitlab.esrf.fr/workflow/ewoks/ewokscore/-/issues/65*

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -32,6 +32,12 @@ test-3.12:
     JUPYTER_PLATFORM_DIRS: "1" # jupyter_client deprecation warning
     PYTEST_WARNINGS: "-W error ${WARNINGS_PY3_12}"
 
+test-3.13:
+  extends: .test-3.13
+  variables:
+    JUPYTER_PLATFORM_DIRS: "1" # jupyter_client deprecation warning
+    PYTEST_WARNINGS: "-W error ${WARNINGS_PY3_12}"
+
 build_sdist:
   extends: .build_sdist
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+New features:
+
+- Support for Python 3.13.
+
 ## 1.0.0
 
 Breaking changes:

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -23,7 +23,7 @@ dependencies = [
     "pyyaml >=5.1",
     "h5py >=2.8",
     "packaging",
-    "ewoksutils >=1.0.0",
+    "ewoksutils >=1.2.0",
     "importlib_metadata;python_version < '3.9'",
 ]
 

--- a/src/ewokscore/events/global_state.py
+++ b/src/ewokscore/events/global_state.py
@@ -7,8 +7,9 @@ import warnings
 from contextlib import contextmanager
 from typing import Dict, Iterable, List, Optional, Tuple
 
-from ewoksutils.logging_utils.asyncwrapper import AsyncHandlerWrapper
 from ewoksutils.logging_utils.cleanup import cleanup_logger
+from ewoksutils.logging_utils.cleanup import protect_logging_state
+from ewoksutils.logging_utils.asyncwrapper import AsyncHandlerWrapper
 
 from .handlers import is_ewoks_event_handler, instantiate_handler
 
@@ -61,11 +62,8 @@ def remove_handler(handler: logging.Handler) -> None:
 
 def cleanup():
     """Pending events will be dropped"""
-    logging._acquireLock()
-    try:
+    with protect_logging_state():
         _cleanup_ewoks_event_logger()
-    finally:
-        logging._releaseLock()
 
 
 def _after_fork_in_child():
@@ -84,15 +82,12 @@ def _ewoks_event_logger(
     # Issue with logging and forking:
     # https://pythonspeed.com/articles/python-multiprocessing/
 
-    logging._acquireLock()
-    try:
+    with protect_logging_state():
         if _ewoks_event_logger_requires_cleanup():
             _cleanup_ewoks_event_logger()
         if _ewoks_event_logger_requires_init():
             _init_ewoks_event_logger(handlers, asynchronous)
         yield logging.getLogger(EWOKS_EVENT_LOGGER_NAME)
-    finally:
-        logging._releaseLock()
 
 
 def _cleanup_ewoks_event_logger():

